serving/purgatorio/select-and-train-machine-learning-models/evaluation-and-fine-tunig/evaluation-and-finetuning.md

```diff
@@ -1,5 +1,169 @@
+
+There are no Golden Rules ([There Is No One Ring to Rule them All](#There-Is-No-One-Ring-to-Rule-them-All)), even though some best practices are recognized to work well in a variety of situations. 
+
+But when you tackle a novel problem, with a custom dataset, it's difficult to choose these hyperparameters right at the first shot. 
+
+For example, imagine you're training a neural network (even a simple one), you can have _a lot_ of these hyperparameters:
+About the topology of the network:
+- the type of layers
+- the number of layers
+- the number of neurons in each layer
+- the activation functions
+
+Or about the optimization phase (the actual training):
+
+- the number of training epochs
+- the batch size 
+- the learning rate
+- the learning rate decay
+- regularization techniques
+
+What a mess! How to get these right?
+
+Often data scientists try to overcome this problem with some techniques, like Grid Search, Random Search, or Evolutionary Algorithms.
+
+Be sure to get through the following resources to understand how to tune the hyperparameters of your model!
+
+- [Introduction to Hyperparameter Tuning](https://community.alteryx.com/t5/Data-Science-Blog/Hyperparameter-Tuning-Black-Magic/ba-p/449289)
+- [Grid Search](https://machinelearningmastery.com/grid-search-hyperparameters-deep-learning-models-python-keras/)
+- [Random Search](https://towardsdatascience.com/feature-selection-by-random-search-in-python-730ffd2912e9)
+- [Cross-validation: evaluating estimator performance - with Scikit-Learn](https://scikit-learn.org/stable/modules/cross_validation.html)
+
+Read also:
+
+- [Practical hyperparameter optimization: Random vs. grid search](https://stats.stackexchange.com/questions/160479/practical-hyperparameter-optimization-random-vs-grid-search)
+- [Practical Guide to Hyperparameters Optimization for Deep Learning Models](https://blog.floydhub.com/guide-to-hyperparameters-search-for-deep-learning-models/)
+- [Hyperparameter tuning with Keras Tuner](https://blog.tensorflow.org/2020/01/hyperparameter-tuning-with-keras-tuner.html)
+
+And more advanced techniques:
+- [Hyp-RL : Hyperparameter Optimization by Reinforcement Learning](https://arxiv.org/abs/1906.11527)
+
+Remember that these techniques are not meant to substitute the Data Scientist knowledge, but to help him in finding the most promising directions! 
+
+With the experience, you'll learn by yourself what are likely good configurations, depending on the problem at hand.
+
+## Conclusions
+
+In this guide, you've learned how to choose the right test set for your evaluation, how to choose a meaningful metric for the problem at hand, and how to evaluate the model against it. 
+
+You've also seen some techniques that help you move in the deep space of the hyperparameters configurations. 
+
+Now, once you're satisfied by the performance fo your model, you're ready to use it in the real world!
```

Comment:
All very good points and resources, I would say please refer to many of the links mentioned in the 
https://github.com/virgili0/Virgilio/pull/89, i.e. https://github.com/virgili0/Virgilio/pull/89#discussion_r406953449, https://github.com/virgili0/Virgilio/pull/89#discussion_r406951997 and others on the PR and also under https://github.com/neomatrix369/awesome-ai-ml-dl